src/middlewared/middlewared/plugins/smart.py

```diff
@@ -296,6 +299,162 @@ class Config:
+                        test_disks_list.append({
+                            'disk': current_disk['name'],
+                            **disk
+                        })
+                        break
+                else:
+                    verrors.add(
+                        'manual_test.disks.identifier',
+                        f'{disk["identifier"]} is not valid. Please provide a valid disk identifier.'
+                    )
+                    continue
+
+                if disk['alert_service_id'] and not (
+                    await self.middleware.call(
+                        'alertservice.query',
+                        [['id', '=', disk['alert_service_id']]]
+                    )
+                ):
+                    verrors.add(
+                        'manual_test.disks.alert_service_id',
+                        f'Please provide a valid alert_service_id for {disk["identifier"]}'
+                    )
+
+                if current_disk['name'] is None or current_disk['name'].startswith('nvd'):
+                    verrors.add(
+                        'manual_test.disks.identifier',
+                        f'Test cannot be performed for {disk["identifier"]} disk. Failed to retrieve name.'
+                    )
+
+                device = devices.get(current_disk['name'])
+                if not device:
+                    verrors.add(
+                        'manual_test.disks.identifier',
+                        f'Test cannot be performed for {disk["identifier"]}. Unable to retrieve disk details.'
+                    )
+
+        verrors.check()
+
+        smartd_pid = await self.middleware.call('smart.smartd_pid')
+
+        return list(
+            await asyncio_map(functools.partial(self.__manual_test, devices, smartd_pid), test_disks_list, 16)
+        )
+
+    async def __manual_test(self, devices, smartd_pid, disk):
+        device = devices.get(disk['disk'])
+        args = await get_smartctl_args(disk['disk'], device)
+
+        proc = await run(
+            list(
+                filter(bool, ['smartctl', '-t', disk['type'].lower(), '-C' if disk['mode'] == 'FOREGROUND' else None])
+            ) + args,
+            check=False, encoding='utf8'
+        )
+
+        output = {}
+        if proc.returncode:
+            output['error'] = proc.stderr
+            syslog.openlog('smartctl', facility=syslog.LOG_DAEMON)
+            syslog.syslog(
+                syslog.LOG_WARNING,
+                f'Self test run for /dev/{disk["disk"]} failed with {proc.returncode} return code'
+            )
+            syslog.closelog()
+        else:
+            time_details = re.findall('test will complete after(.*)', proc.stdout, re.IGNORECASE)
+            if not time_details:
+                output['error'] = f'Failed to parse smartctl self test details for {disk["identifier"]}.'
+            else:
+                output['Expected Result Time'] = time_details[0].strip()
```

Comment:
Does it always produce the right time? Can't it say that it will be complete in 2 hours and finish in 5 minutes instead?

Reply:
I have been doing this for a while now and have tested this on 11.3/12, it doesn't change as far as i have seen. I can't find a reference which would explicitly say that indeed it doesn't. Please advise if you think there can be a fluctuation in the format ( It is dependent on the locale i think ).

Reply:
It's not the format that bothers me, it's the value, is it right to check only in 2 hours if it says that test will complete in 2 hours? What if it fails in 2 minutes instead?